File: vol2/ch18/LeetCode2608.py
class Solution:
    def findShortestCycle(self, n: int, edges: List[List[int]]) -> int:
        self.INF=0x3f3f3f3f
        self.adj=defaultdict(set)
        for u,v in edges:                               #创建邻接表adj
            self.adj[u].add(v)
            self.adj[v].add(u)
        ans=self.INF
        for u in range(0,n):
            for v in self.adj[u]:
                ans=min(ans,self.bfs(n,u,v)+1)
        return -1 if ans==self.INF else ans
    def bfs(self,n,u,v):
        dist=[self.INF]*n
        qu=deque()
        qu.append(u)
        dist[u]=0
        while qu:
            i=qu.popleft()
            for j in self.adj[i]:
                if dist[j]==self.INF:
                    if i==u and j==v:continue;
                    qu.append(j)
                    dist[j]=dist[i]+1
        return dist[v]

File: vol2/ch18/LeetCode2608.cpp
class Solution {
    const int INF=0x3f3f3f3f;
    vector<vector<int>> adj;
public:
    int findShortestCycle(int n,vector<vector<int>>& edges) {
        adj=vector<vector<int>>(n);
        for(auto e:edges) {             //创建邻接表adj
            adj[e[0]].push_back(e[1]);
            adj[e[1]].push_back(e[0]);
        }
        int ans=INF;
        for (int u=0;u<n;u++) {
            for (int v:adj[u]) {
                ans=min(ans,bfs(n,u,v)+1);
            }
        }
        return ans==INF?-1:ans;
    }
    int bfs(int n,int u,int v) {
        vector<int> dist(n,INF);
        queue<int> qu;
        qu.push(u);
        dist[u]=0;
        while (!qu.empty()) {
            int i=qu.front(); qu.pop();
            for (int j:adj[i]) {
                if (dist[j]==INF) {
                    if (i==u && j==v) continue;
                    qu.push(j);
                    dist[j]=dist[i]+1;
                }
            }
        }
        return dist[v];
    }
};

